### ISSUE TITLE
Failed to download cover

### ISSUE THREAD

**Issue by idMysteries** [dipu-bd/lightnovel-crawler#1669]
@dipu-bd failed to get cover again... https://github.com/dipu-bd/lightnovel-crawler/pull/1662

https://puretl.com/assassin-farmer
http://static1.squarespace.com/static/61d631c6bb5b2a1ff76d5426/t/62253455e3cd1828c1d8bb25/1646605397532/unsplash-image-rdVLN3UFmpU.jpg?format=1500w

_Originally posted by @idMysteries in https://github.com/dipu-bd/lightnovel-crawler/issues/1635#issuecomment-1264308157_

**Comment by idMysteries** [comment]
@dipu-bd In my browser, http redirects to https. Maybe that's the problem?

**Comment by dipu-bd** [comment]
unsplash was causing problem. I created a random cover image generator

**Comment by idMysteries** [comment]
NO IT'S NOT unsplash

**Comment by dipu-bd** [comment]
I tested in several VPNs and found out in some of them unsplash image url does not work. If I want to guarantee having a cover image, we can not actually rely on any online sources for it. That is why a offline image generator was required.

Using custom image generator also solves this issue. If you have some better proposal please say.

**Comment by idMysteries** [comment]
@dipu-bd IT'S NOT UNSPLASH!!!

**Comment by dipu-bd** [comment]
Maybe I am not getting what your issue is.  Explain more about what the issue is, how to reproduce it. There is a bug issue template for it. 

This issue title says about novel cover. If you are experiencing another issue with content image download, create another ticket for it specifying the source name.

Since the novel cover issue is solved, I closed this ticket.

**Comment by idMysteries** [comment]
![image](https://user-images.githubusercontent.com/11484976/193451017-ab8b031a-0fb8-4c0e-87e3-767451464b8d.png)
@dipu-db there was no problem with the generator here.

**Comment by dipu-bd** [comment]
> @dipu-db there was no problem with the generator here.

I have tested and found out that there is an issue.  Let me tell you in DETAILS what I discovered and what I fixed.

The issue you are talking about is actually a source specific issue. Cloudflare was blocking images to be downloaded in some of the sources. Other sources not having cloudflare are working fine. e.g.: https://novelfull.com/sovereign-of-the-karmic-system.html

In case a cover image download fails, to not let cover go empty, a random image was being downloaded from unsplash. I tested and found out even that random image was failing to download in some cases and cover goes empty.

To not let the cover image be empty, an offline image generator can be used. Now if the original cover image fails to download, the generator will create a random cover image instead of getting a random image from unsplash.

In future, create source specific issues using the template: https://github.com/dipu-bd/lightnovel-crawler/issues/new?assignees=&labels=source-issue&template=3_source-issue.md&title=Fix+this+source

And do not use f-words without understanding the main issue.
